README.markdown
Yahoo! Social SDK - PHP
=======================

Find documentation and support on Yahoo! Developer Network: http://developer.yahoo.com

 * Yahoo! Application Platform - http://developer.yahoo.com/yap/
 * Yahoo! Social APIs - http://developer.yahoo.com/social/
 * Yahoo! Query Language - http://developer.yahoo.com/yql/

Hosted on GitHub: http://github.com/yahoo/yos-social-php/tree/master


License
=======

@copyright: Copyrights for code authored by Yahoo! Inc. is licensed under the following terms:
@license:   BSD Open Source License

Yahoo! Social SDK
Software License Agreement (BSD License)
Copyright (c) 2009, Yahoo! Inc.
All rights reserved.

Redistribution and use of this software in source and binary forms, with
or without modification, are permitted provided that the following
conditions are met:

* Redistributions of source code must retain the above
  copyright notice, this list of conditions and the
  following disclaimer.

* Redistributions in binary form must reproduce the above
  copyright notice, this list of conditions and the
  following disclaimer in the documentation and/or other
  materials provided with the distribution.

* Neither the name of Yahoo! Inc. nor the names of its
  contributors may be used to endorse or promote products
  derived from this software without specific prior
  written permission of Yahoo! Inc.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


The Yahoo! Social PHP SDK code is subject to the BSD license, see the LICENSE file.


Requirements
============

The following dependencies are bundled with the Yahoo! PHP SDK, but are under
terms of a separate license. See the bundled LICENSE files for more information:

 * OAuth - http://code.google.com/p/oauth
 * JSON -  http://pear.php.net/Services_JSON


Install
=======

After downloading and unpackaging the release, copy the contents of yosdk/lib
to a directory that is accessible via the PHP require_once method. In
your PHP scripts, require the Yahoo.inc file:

    require_once("Yahoo.inc");

## Debugging with the YOS SDK

The YOS SDK provides some methods that make debugging easier. Also, since the YOS
SDK was designed for the greatest compatibility across PHP version, it is advised
to suppress notices if using PHP 5.x:

    // debug settings
    error_reporting(E_ALL | E_NOTICE); # do not show notices as library is php4 compatable
    ini_set('display_errors', true);

    YahooLogger::setDebug(true);
    YahooLogger::setDebugDestination('LOG');

By default, OAuth credentials (request + access tokens) are stored in PHP sessions.
PHP sessions can easily be configured to work from a local filesystem, a database, or
a memcache instance. The sample apps provide working examples of filesystem storage
and memcache storage.

## Filesystem Session Storage (default)

    // use local filesystem to store oauth credentials via php native sessions
    ini_set('session.save_handler', 'files');
    session_save_path('/tmp/');
    session_start();

## Memcache Session Storage

    // use memcache to store oauth credentials via php native sessions
    ini_set('session.save_handler', 'memcache');
    session_save_path('tcp://localhost:11211?persistent=1&weight=1&timeout=1&retry_interval=15');
    session_start();

## Cookie Based Session Storage (original)

The original default storage backend was based on storing access tokens in a plain text cookie. This is bad for security, and the default has since been changed (breaks backwards compatibility). In order to use the original
session store, you must instantiate 'CookieSessionStore' and pass into all methods of YahooSession:

    $sessionStore = new CookieSessionStore();

    YahooSession::hasSession($consumerKey, $consumerSecret, $applicationId, $sessionStore, $verifier)
    YahooSession::requireSession($consumerKey, $consumerSecret, $applicationId, $callback, $sessionStore, $verifier)
    YahooSession::initSession($consumerKey, $consumerSecret, $applicationId, $redirect, $callback, $sessionStore, $verifier)

    YahooSession::createAuthorizationUrl($consumerKey, $consumerSecret, $callback, $sessionStore)
    YahooSession::redirectForAuthorization($consumerKey, $consumerSecret, $callback, $sessionStore)
    YahooSession::clearSession($sessionStore)
    YahooSession::checkSession($sessionType, $sessionStore)


Examples
========

In the yosdk/sample directory you'll find sampleapp.php, a sample application
that will get you started. Put the file in your hosting directory and fill
in the consumer key, consumer key secret and application ID. When you're done,
point your browser at the URL. You will be directed to verify access to the
application and then you will be redirected back to your application.

In the yosdk/sample directory you'll find delicious.php, a sample yahoo open
application that shows how to use YQL to make web service request.

Make sure to update the OAuth credentials in each bundled example:

    define('OAUTH_CONSUMER_KEY', '###');
    define('OAUTH_CONSUMER_SECRET', '###');
    define('OAUTH_DOMAIN', '###');
    define('OAUTH_APP_ID', '###');


## Fetching YQL:

    $session = YahooSession::requireSession(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET, OAUTH_APP_ID);
    $user = $session->getSessionedUser();

    var_dump($session->query('select * from social.connections where owner_guid=me'));


## Fetching Social Data:

    $session = YahooSession::requireSession(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET, OAUTH_APP_ID);
    $user = $session->getSessionedUser();

    var_dump($user->getProfile());
    var_dump($user->getContacts());


## Setting YAP Small View:

    $session = YahooSession::requireSession(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET, OAUTH_APP_ID);
    $user = $session->getSessionedUser();

    $content = '<h1><yml:name> customized small view</h1>';
    var_dump($user->setSmallView($content));


## Adding a Contact:
    
    $contact_fields = array();
    $contact_fields[] = array('type' => 'email', 'value' => 'me@domain.com');
    $contact_fields[] = array('type' => 'name', 'value' => array('givenName'=> 'John', 'familyName' => 'Doe'));
    $contact = array('fields' => $contact_fields);

    var_dump($user->addContact($contact));

Tests
=====

The Yahoo! PHP SDK comes with a test suite to validate functionality. The tests also
show functional examples and results. To run the test suite, simply execute the test suite:

    php phpunit test/OAuthTest.php


CHANGELOG
Yahoo! Social SDK - PHP - CHANGELOG
===================================

v1.0 - initial stable release


test/unit/oauth/OAuthUtilTest.php
<?php

require_once dirname(__FILE__).'/../../../lib/OAuth.php';
require_once 'PHPUnit/Framework.php';

/**
 * Tests of OAuthUtil
 */
class OAuthUtilTest extends PHPUnit_Framework_TestCase {
  public function testUrlencode() {
    // Tests taken from
    // http://wiki.oauth.net/TestCases ("Parameter Encoding")
    $this->assertEquals('abcABC123', OAuthUtil::urlencode_rfc3986('abcABC123'));
    $this->assertEquals('-._~',      OAuthUtil::urlencode_rfc3986('-._~'));
    $this->assertEquals('%25',       OAuthUtil::urlencode_rfc3986('%'));
    $this->assertEquals('%2B',       OAuthUtil::urlencode_rfc3986('+'));
    $this->assertEquals('%0A',       OAuthUtil::urlencode_rfc3986("\n"));
    $this->assertEquals('%20',       OAuthUtil::urlencode_rfc3986(' '));
    $this->assertEquals('%7F',       OAuthUtil::urlencode_rfc3986("\x7F"));
    //$this->assertEquals('%C2%80',    OAuthUtil::urlencode_rfc3986("\x00\x80"));
    //$this->assertEquals('%E3%80%81', OAuthUtil::urlencode_rfc3986("\x30\x01"));

    // Last two checks disabled because of lack of UTF-8 support, or lack
    // of knowledge from me (morten.fangel) on how to use it properly..

    // A few tests to ensure code-coverage
    $this->assertEquals( '', OAuthUtil::urlencode_rfc3986(NULL));
    $this->assertEquals( '', OAuthUtil::urlencode_rfc3986(new stdClass()));
  }

  public function testUrldecode() {
    // Tests taken from
    // http://wiki.oauth.net/TestCases ("Parameter Encoding")
    $this->assertEquals('abcABC123', OAuthUtil::urldecode_rfc3986('abcABC123'));
    $this->assertEquals('-._~',      OAuthUtil::urldecode_rfc3986('-._~'));
    $this->assertEquals('%',         OAuthUtil::urldecode_rfc3986('%25'));
    $this->assertEquals('+',         OAuthUtil::urldecode_rfc3986('%2B'));
    $this->assertEquals("\n",        OAuthUtil::urldecode_rfc3986('%0A'));
    $this->assertEquals(' ',         OAuthUtil::urldecode_rfc3986('%20'));
    $this->assertEquals("\x7F",      OAuthUtil::urldecode_rfc3986('%7F'));
    //$this->assertEquals("\x00\x80",  OAuthUtil::urldecode_rfc3986('%C2%80'));
    //$this->assertEquals("\x30\x01",  OAuthUtil::urldecode_rfc3986('%E3%80%81'));

    // Last two checks disabled because of lack of UTF-8 support, or lack
    // of knowledge from me (morten.fangel) on how to use it properly..
  }

  public function testParseParameter() {
    // Tests taken from
    // http://wiki.oauth.net/TestCases ("Normalize Request Parameters")

    $this->assertEquals(
      array('name'=>''),
      OAuthUtil::parse_parameters('name')
    );
    $this->assertEquals(
      array('a'=>'b'),
      OAuthUtil::parse_parameters('a=b')
    );
    $this->assertEquals(
      array('a'=>'b','c'=>'d'),
      OAuthUtil::parse_parameters('a=b&c=d')
    );
    $this->assertEquals(
      array('a'=>array('x!y','x y')),
      OAuthUtil::parse_parameters('a=x!y&a=x+y')
    );
    $this->assertEquals(
      array('x!y'=>'a', 'x' =>'a'),
      OAuthUtil::parse_parameters('x!y=a&x=a')
    );
  }

  public function testBuildHttpQuery() {
    // Tests taken from
    // http://wiki.oauth.net/TestCases ("Normalize Request Parameters")
    $this->assertEquals(
        'name=',
      OAuthUtil::build_http_query(array('name'=>''))
    );
    $this->assertEquals(
      'a=b',
      OAuthUtil::build_http_query(array('a'=>'b'))
    );
    $this->assertEquals(
      'a=b&c=d',
      OAuthUtil::build_http_query(array('a'=>'b','c'=>'d'))
    );
    $this->assertEquals(
      'a=x%20y&a=x%21y',
      OAuthUtil::build_http_query(array('a'=>array('x!y','x y')))
    );
    $this->assertEquals(
      'x=a&x%21y=a',
      OAuthUtil::build_http_query(array('x!y'=>'a', 'x' =>'a'))
    );

    // Test taken from the Spec 9.1.1
    $this->assertEquals(
      'a=1&c=hi%20there&f=25&f=50&f=a&z=p&z=t',
      OAuthUtil::build_http_query(array('a'=>'1', 'c' =>'hi there', 'f'=>array(25, 50, 'a'), 'z'=>array('p','t')))
    );
  }

  public function testSplitHeader() {
    $this->assertEquals(
      array('oauth_foo'=>'bar','oauth_baz'=>'bla,rgh'),
      OAuthUtil::split_header('OAuth realm="",oauth_foo=bar,oauth_baz="bla,rgh"')
    );
    $this->assertEquals(
      array(),
      OAuthUtil::split_header('OAuth realm="",foo=bar,baz="bla,rgh"')
    );
    $this->assertEquals(
      array('foo'=>'bar', 'baz'=>'bla,rgh'),
      OAuthUtil::split_header('OAuth realm="",foo=bar,baz="bla,rgh"', false)
    );
    $this->assertEquals(
      array('oauth_foo' => 'hi there'),
      OAuthUtil::split_header('OAuth realm="",oauth_foo=hi+there,foo=bar,baz="bla,rgh"')
    );

  }

  public function testGetHeaders() {
    if (function_exists('apache_request_headers')) {
      $this->markTestSkipped('We assume the apache module is well tested. Since this module is present, no need testing our suplement');
    }

    $_SERVER['HTTP_HOST'] = 'foo';
    $_SERVER['HTTP_X_WHATEVER'] = 'bar';
    $this->assertEquals( array('Host'=>'foo', 'X-Whatever'=>'bar'), OAuthUtil::get_headers() );
  }
}


test/unit/oauth/OAuthTokenTest.php
<?php

require_once dirname(__FILE__).'/../../../lib/OAuth.php';
require_once 'PHPUnit/Framework.php';

class OAuthTokenTest extends PHPUnit_Framework_TestCase {
  public function testSerialize() {
    $token = new OAuthToken('token', 'secret');
    $this->assertEquals('oauth_token=token&oauth_token_secret=secret', $token->to_string());

    $token = new OAuthToken('token&', 'secret%');
    $this->assertEquals('oauth_token=token%26&oauth_token_secret=secret%25', $token->to_string());
  }
  public function testConvertToString() {
    $token = new OAuthToken('token', 'secret');
    $this->assertEquals('oauth_token=token&oauth_token_secret=secret', (string) $token);

    $token = new OAuthToken('token&', 'secret%');
    $this->assertEquals('oauth_token=token%26&oauth_token_secret=secret%25', (string) $token);
  }
}

test/unit/oauth/OAuthConsumerTest.php
<?php

require_once dirname(__FILE__).'/../../../lib/OAuth.php';
require_once 'PHPUnit/Framework.php';

class OAuthConsumerTest extends PHPUnit_Framework_TestCase {
  public function testConvertToString() {
    $consumer = new OAuthConsumer('key', 'secret');
    $this->assertEquals('OAuthConsumer[key=key,secret=secret]', (string) $consumer);
  }
}

sample/sampleapp.php
<?php

// Include the YOS library.
require dirname(__FILE__).'/../lib/Yahoo.inc';

// debug settings
error_reporting(E_ALL | E_NOTICE); # do not show notices as library is php4 compatable
ini_set('display_errors', true);
YahooLogger::setDebug(true);
YahooLogger::setDebugDestination('LOG');

// use memcache to store oauth credentials via php native sessions
ini_set('session.save_handler', 'files');
session_save_path('/tmp/');
session_start();

// Make sure you obtain application keys before continuing by visiting:
// https://developer.yahoo.com/dashboard/createKey.html

define('OAUTH_CONSUMER_KEY', '###');
define('OAUTH_CONSUMER_SECRET', '###');
define('OAUTH_DOMAIN', '###');
define('OAUTH_APP_ID', '###');

if(array_key_exists("logout", $_GET)) {
  // if a session exists and the logout flag is detected
  // clear the session tokens and reload the page.
  YahooSession::clearSession();
  header("Location: sampleapp.php");
}

// check for the existance of a session.
// this will determine if we need to show a pop-up and fetch the auth url,
// or fetch the user's social data.
$hasSession = YahooSession::hasSession(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET, OAUTH_APP_ID);

if($hasSession == FALSE) {
  // create the callback url,
  $callback = YahooUtil::current_url()."?in_popup";

  // pass the credentials to get an auth url.
  // this URL will be used for the pop-up.
  $auth_url = YahooSession::createAuthorizationUrl(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET, $callback);
}
else {
  // pass the credentials to initiate a session
  $session = YahooSession::requireSession(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET, OAUTH_APP_ID);

  // if the in_popup flag is detected,
  // the pop-up has loaded the callback_url and we can close this window.
  if(array_key_exists("in_popup", $_GET)) {
    close_popup();
    exit;
  }

  // if a session is initialized, fetch the user's profile information
  if($session) {
    // Get the currently sessioned user.
    $user = $session->getSessionedUser();

    // Load the profile for the current user.
    $profile = $user->getProfile();
  }
}

/**
 * Helper method to close the pop-up window via javascript.
 */
function close_popup() {
?>
<script type="text/javascript">
  window.close();
</script>
<?php
}
?>
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN">
<html>
  <head>
    <title>YOS Social Platform Sample Application</title>

    <!-- Combo-handled YUI JS files: -->
    <script type="text/javascript" src="http://yui.yahooapis.com/combo?2.7.0/build/yahoo-dom-event/yahoo-dom-event.js"></script>
    <script type="text/javascript" src="popupmanager.js"></script>

    <!-- Combo-handled YUI CSS files: -->
    <link rel="stylesheet" type="text/css" href="http://yui.yahooapis.com/combo?2.7.0/build/reset-fonts-grids/reset-fonts-grids.css&2.7.0/build/base/base-min.css">
  </head>
  <body>
    <?php
      if($hasSession == FALSE) {
        // if a session does not exist, output the
        // login / share button linked to the auth_url.
        echo sprintf("<a href=\"%s\" id=\"yloginLink\"><img src=\"http://l.yimg.com/a/i/ydn/social/updt-spurp.png\"></a>\n", $auth_url);
      }
      else if($hasSession && $profile) {
        // if a session does exist and the profile data was
        // fetched without error, print out a simple usercard.
        echo sprintf("<img src=\"%s\"/><p><h2>Hi <a href=\"%s\" target=\"_blank\">%s!</a></h2></p>\n", $profile->image->imageUrl, $profile->profileUrl, $profile->nickname);

        if($profile->status->message != "") {
          $statusDate = date('F j, y, g:i a', strtotime($profile->status->lastStatusModified));
          echo sprintf("<p><strong>&#8220;</strong>%s<strong>&#8221;</strong> on %s</p>", $profile->status->message, $statusDate);
        }

        echo "<p><a href=\"?logout\">Logout</a></p>";
      }
    ?>
    <script type="text/javascript">
      var Event = YAHOO.util.Event;
      var _gel = function(el) {return document.getElementById(el)};

      function handleDOMReady() {
        if(_gel("yloginLink")) {
          Event.addListener("yloginLink", "click", handleLoginClick);
        }
      }

      function handleLoginClick(event) {
        // block the url from opening like normal
        Event.preventDefault(event);

        // open pop-up using the auth_url
        var auth_url = _gel("yloginLink").href;
        PopupManager.open(auth_url,600,435);
      }

      Event.onDOMReady(handleDOMReady);
    </script>
  </body>
</html>

sample/delicious.php
<?php

/**
 * Using YQL Open Tables to create Yahoo! Open Application - Delicious
 *
 * http://developer.yahoo.com/devtool
 * http://developer.yahoo.com/yql/console
 */

// Include the YOS library.
require dirname(__FILE__).'/../lib/Yahoo.inc';

// debug settings
error_reporting(E_ALL | E_NOTICE); # do not show notices as library is php4 compatable
ini_set('display_errors', true);
YahooLogger::setDebug(true);
YahooLogger::setDebugDestination('LOG');

// use memcache to store oauth credentials via php native sessions
ini_set('session.save_handler', 'memcache');
session_save_path('tcp://localhost:11211?persistent=1&weight=1&timeout=1&retry_interval=15');
session_start();

// Make sure you obtain application keys before continuing by visiting:
// https://developer.yahoo.com/dashboard/createKey.html

// create a Yahoo! Open Application - http://developer.yahoo.com/dashboard

define('OAUTH_CONSUMER_KEY', '###');
define('OAUTH_CONSUMER_SECRET', '###');
define('OAUTH_DOMAIN', '###');
define('OAUTH_APP_ID', '###');


// oauth dance if not authenticated
$session = YahooSession::requireSession(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET, OAUTH_APP_ID);

$header   = '<h2><img src="http://delicious.com/favicon.ico" title="Delicious" width="16" height="16" />Delicious // social bookmarking</h2>';
$content
= '<div id="bookmarks">';

// if user is logged in and oauth is valid
if(is_object($session))
{
  // get oauthed user guid + profile
  $user = $session->getSessionedUser();

  // load y! profile data
  $profile = $user->getProfile();

  // get yap app instance for yql / small view
  $application = new YahooApplication(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET);

  // delicious yql
  $yql = "use 'http://www.javarants.com/delicious/delicious.feeds.xml' as delicious.feeds;";
  $yql .= "select * from delicious.feeds";

  if(isset($_REQUEST['username']))
  {
    // validate delicious user and set username in datastore
    $username = htmlspecialchars($_REQUEST['username']);
  }
  else
  {
    // fetch user data store
    $username = apc_fetch('user_'.$user->guid);
  }

  if($username)
  {
    // if user has specified a delicious username
    $yql .= " where username='".$username."';";
  }

  $data = $application->query($yql);
  if(is_object($data) && isset($data->query->results->item))
  {
    // build small view content (my yahoo)
    $content .= '<ul class="delicious">';
      foreach($data->query->results->item as $bookmark):
        $content .= '<li><a href="'.$bookmark->link.'">'.$bookmark->title.'</a></li>';
      endforeach;
    $content .= '</ul>';

    if($username)
    {
      if(isset($_REQUEST['username']))
      {
        //  username is valid (yql returned results), so save for future requests
        apc_store('user_'.$user->guid, $username);
      }

      // update small view for user bookmarks
      $application->setSmallView($user->guid, $header.$content);
    }
  }
  else
  {
    $content .= '<ul><li><a href="http://delicious.com/popular">Invalid username...</a></li></ul>';
  }
}
$content  .= '</div>';

/*
 // available open tables

  use 'http://www.javarants.com/delicious/delicious.feeds.popular.xml' as delicious.feeds.popular;
  use 'http://www.javarants.com/delicious/delicious.feeds.xml' as delicious.feeds;

  use 'http://www.javarants.com/friendfeed/friendfeed.feeds.xml' as friendfeed.feeds;
  use 'http://www.javarants.com/friendfeed/friendfeed.home.xml' as friendfeed.home;
  use 'http://www.javarants.com/friendfeed/friendfeed.profile.xml' as friendfeed.profile;
  use 'http://www.javarants.com/friendfeed/friendfeed.rooms.xml' as friendfeed.rooms;
  use 'http://www.javarants.com/friendfeed/friendfeed.services.xml' as friendfeed.services;
  use 'http://www.javarants.com/friendfeed/friendfeed.updates.xml' as friendfeed.updates;

  use 'http://www.javarants.com/nyt/nyt.article.search.xml' as nyt.article.search;
  use 'http://www.javarants.com/nyt/nyt.bestsellers.history.xml' as nyt.bestsellers.history;
  use 'http://www.javarants.com/nyt/nyt.bestsellers.search.xml' as nyt.bestsellers.search;
  use 'http://www.javarants.com/nyt/nyt.bestsellers.xml' as nyt.bestsellers;
  use 'http://www.javarants.com/nyt/nyt.movies.critics.xml' as nyt.movies.critics;
  use 'http://www.javarants.com/nyt/nyt.movies.picks.xml' as nyt.movies.picks;
  use 'http://www.javarants.com/nyt/nyt.movies.reviews.xml' as nyt.movies.reviews;

  use 'http://www.javarants.com/shopping/shopping.product.search.xml' as shopping.product.search;

  use 'http://www.javarants.com/weather/weather.local.xml' as weather.local;
  use 'http://www.javarants.com/weather/weather.search.xml' as weather.search;

  use 'http://www.javarants.com/dopplr/dopplr.auth.xml' as dopplr.auth;
  use 'http://www.javarants.com/dopplr/dopplr.city.info.xml' as dopplr.city.info;
  use 'http://www.javarants.com/dopplr/dopplr.futuretrips.info.xml' as dopplr.futuretrips.info;
  use 'http://www.javarants.com/dopplr/dopplr.traveller.fellows.xml' as dopplr.traveller.fellows;
  use 'http://www.javarants.com/dopplr/dopplr.traveller.info.xml' as dopplr.traveller.info;
  use 'http://www.javarants.com/dopplr/dopplr.traveller.travelling.xml' as dopplr.traveller.travelling;
  use 'http://www.javarants.com/dopplr/dopplr.trips.info.xml' as dopplr.trips.info;

  use 'http://www.javarants.com/twitter/twitter.user.timeline.xml' as twitter.user.timeline;
  use 'http://www.javarants.com/twitter/twitter.user.profile.xml' as twitter.user.profile;
  use 'http://www.javarants.com/twitter/twitter.user.status.xml' as twitter.user.status;

  use 'http://www.javarants.com/wesabe/wesabe.tags.xml' as wesabe.tags;

  use 'http://www.javarants.com/whitepages/whitepages.search.xml' as whitepages.search;
  use 'http://www.javarants.com/whitepages/whitepages.reverse.xml' as whitepages.reverse;

  use 'http://www.javarants.com/github/github.repo.commits.xml' as github.repo.commits;
  use 'http://www.javarants.com/github/github.user.repos.xml' as github.user.repos;
  use 'http://www.javarants.com/github/github.user.info.xml' as github.user.info;

  use 'http://www.javarants.com/zillow/zillow.search.xml' as zillow.search;
*/

// set content type + language
// header("Content-Type: text/html; charset=utf-8");
// header("Content-Language: en");
/*
echo '<?xml version="1.0" encoding="utf-8"?>';
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" dir="ltr">
  <head profile="http://gmpg.org/xfn/11">
    <title>Delicious // social bookmarking</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta http-equiv="Content-Language" content="en" />
    <link rel="stylesheet" type="text/css" media="screen" href="http://yui.yahooapis.com/combo?3.0.0pr2/build/cssreset/reset-min.css&amp;3.0.0pr2/build/cssfonts/fonts-min.css&amp;3.0.0pr2/build/cssgrids/grids-min.css&amp;3.0.0pr2/build/cssbase/base-min.css&amp;3.0.0pr2/build/widget/assets/skins/sam/widget.css&amp;3.0.0pr2/build/widget/assets/skins/sam/widget-stack.css&amp;3.0.0pr2/build/overlay/assets/skins/sam/overlay.css" />
    <link rel="stylesheet" type="text/css" media="screen" href="common.css" />
    <link rel="shortcut icon" href="http://delicious.com/favicon.ico" />
    <body>

    <div id="itw" class="yui-d0 yui-skin-sam">
      <?php echo $content; ?>
    </div>

    <script type="text/javascript" src="http://yui.yahooapis.com/combo?3.0.0pr2/build/yui/yui-min.js&amp;3.0.0pr2/build/oop/oop-min.js&amp;3.0.0pr2/build/event/event-min.js&amp;3.0.0pr2/build/dom/dom-min.js&amp;3.0.0pr2/build/node/node-min.js&amp;3.0.0pr2/build/attribute/attribute-min.js&amp;3.0.0pr2/build/base/base-min.js&amp;3.0.0pr2/build/anim/anim-min.js&amp;3.0.0pr2/build/cookie/cookie-min.js&amp;3.0.0pr2/build/io/io-min.js&amp;3.0.0pr2/build/json/json-min.js&amp;3.0.0pr2/build/dump/dump-min.js&amp;3.0.0pr2/build/substitute/substitute-min.js&amp;3.0.0pr2/build/classnamemanager/classnamemanager-min.js&amp;3.0.0pr2/build/widget/widget-min.js&amp;3.0.0pr2/build/widget/widget-position-min.js&amp;3.0.0pr2/build/widget/widget-position-ext-min.js&amp;3.0.0pr2/build/widget/widget-stack-min.js&amp;3.0.0pr2/build/widget/widget-stdmod-min.js&amp;3.0.0pr2/build/overlay/overlay-min.js"></script>
    <script type="text/javascript" src="common.js"></script>
    </body>
</html>
*/
?>
<style type="text/css">
.yui-d0,.yui-d1,.yui-d1f,.yui-d2,.yui-d2f,.yui-d3,.yui-d3f{margin:auto;text-align:left;width:57.69em;}.yui-t1,.yui-t2,.yui-t3,.yui-t4,.yui-t5,.yui-t6{margin:auto;text-align:left;width:100%;}.yui-d0{margin:auto 10px;width:auto;}.yui-d0f{width:100%;}.yui-d2{width:73.076em;}.yui-d2f{width:950px;}.yui-d3{width:74.923em;}.yui-d3f{width:974px;}.yui-b{position:relative;}.yui-main .yui-b{position:static;}.yui-main{width:100%;}.yui-t1 .yui-main,.yui-t2 .yui-main,.yui-t3 .yui-main{float:right;margin-left:-25em;}.yui-t4 .yui-main,.yui-t5 .yui-main,.yui-t6 .yui-main{float:left;margin-right:-25em;}.yui-t1 .yui-b{float:left;width:12.30769em;}.yui-t1 .yui-main .yui-b{margin-left:13.30769em;}.yui-t2 .yui-b{float:left;width:13.84615em;}.yui-t2 .yui-main .yui-b{margin-left:14.84615em;}.yui-t3 .yui-b{float:left;width:23.0769em;}.yui-t3 .yui-main .yui-b{margin-left:24.0769em;}.yui-t4 .yui-b{float:right;width:13.8456em;}.yui-t4 .yui-main .yui-b{margin-right:14.8456em;}.yui-t5 .yui-b{float:right;width:18.4615em;}.yui-t5 .yui-main .yui-b{margin-right:19.4615em;}.yui-t6 .yui-b{float:right;width:23.0769em;}.yui-t6 .yui-main .yui-b{margin-right:24.0769em;}.yui-main .yui-b{float:none;width:auto;}.yui-gb .yui-u,.yui-g .yui-gb .yui-u,.yui-gb .yui-g,.yui-gb .yui-gb,.yui-gb .yui-gc,.yui-gb .yui-gd,.yui-gb .yui-ge,.yui-gb .yui-gf,.yui-gc .yui-u,.yui-gc .yui-g,.yui-gd .yui-u{float:left;}.yui-g .yui-u,.yui-g .yui-g,.yui-g .yui-gb,.yui-g .yui-gc,.yui-g .yui-gd,.yui-g .yui-ge,.yui-g .yui-gf,.yui-gc .yui-u,.yui-gd .yui-g,.yui-g .yui-gc .yui-u,.yui-ge .yui-u,.yui-ge .yui-g,.yui-gf .yui-g,.yui-gf .yui-u{float:right;}.yui-g div.first,.yui-gb div.first,.yui-gc div.first,.yui-gd div.first,.yui-ge div.first,.yui-gf div.first,.yui-g .yui-gc div.first,.yui-g .yui-ge div.first,.yui-gc div.first div.first{float:left;}.yui-g .yui-u,.yui-g .yui-g,.yui-g .yui-gb,.yui-g .yui-gc,.yui-g .yui-gd,.yui-g .yui-ge,.yui-g .yui-gf{width:49.1%;}.yui-gb .yui-u,.yui-g .yui-gb .yui-u,.yui-gb .yui-g,.yui-gb .yui-gb,.yui-gb .yui-gc,.yui-gb .yui-gd,.yui-gb .yui-ge,.yui-gb .yui-gf,.yui-gc .yui-u,.yui-gc .yui-g,.yui-gd .yui-u{width:32%;margin-left:2.0%;}.yui-gc div.first,.yui-gd .yui-u{width:66%;}.yui-gd div.first{width:32%;}.yui-ge div.first,.yui-gf .yui-u{width:74.2%;}.yui-ge .yui-u,.yui-gf div.first{width:24%;}.yui-g .yui-gb div.first,.yui-gb div.first,.yui-gc div.first,.yui-gd div.first{margin-left:0;}.yui-g .yui-g .yui-u,.yui-gb .yui-g .yui-u,.yui-gc .yui-g .yui-u,.yui-gd .yui-g .yui-u,.yui-ge .yui-g .yui-u,.yui-gf .yui-g .yui-u{width:49%;}.yui-g .yui-gc div.first,.yui-gd .yui-g{width:66%;}.yui-g .yui-gc .yui-u,.yui-gb .yui-gc .yui-u{width:32%;margin-right:0;}.yui-gb .yui-gc div.first{width:66%;}.yui-gb .yui-ge .yui-u,.yui-gb .yui-gf .yui-u{margin:0;}.yui-gb .yui-gd div.first{width:32%;}.yui-ge .yui-g{width:24%;}.yui-gf .yui-g{width:74.2%;}.yui-gb .yui-ge div.yui-u,.yui-gb .yui-gf div.yui-u{float:right;}.yui-gb .yui-ge div.first,.yui-gb .yui-gf div.first{float:left;}.yui-ge div.first .yui-gd .yui-u{width:65%;}.yui-ge div.first .yui-gd div.first{width:32%;}
h1{font-size:138.5%;}h2{font-size:123.1%;}h3{font-size:108%;}h1,h2,h3{margin:1em 0;}h1,h2,h3,h4,h5,h6,strong{font-weight:bold;}abbr,acronym{border-bottom:1px dotted #000;cursor:help;}em{font-style:italic;}blockquote,ul,ol,dl{margin:1em;}ol,ul,dl{margin-left:2em;}ol li{list-style:decimal outside;}ul li{list-style:disc outside;}dl dd{margin-left:1em;}th,td{border:1px solid #000;padding:.5em;}th{font-weight:bold;text-align:center;}caption{margin-bottom:.5em;text-align:center;}p,fieldset,table,pre{margin-bottom:1em;}textarea{width:12.25em;}

#itw {
  text-align: left;
  font-family: Tahoma, Georgia;
  background-color: #fff;
  color: #000;
}


img {
  margin: 4px;
}

a {
  text-decoration: none;
  font-weight: bold;
  color: #000;
}

a:hover {
  color: #000;
  text-decoration:underline;
}

div.delicious {
  padding: 0 0 0 0;
  margin: 0 0 0 0;
}

div.delicious ul {
  padding: 0 0 0 0;
  margin: 0 0 0 0;
}

div.delicious ul li {
  list-style-type: none;
  padding: 0 0 0 0;
  margin: 0 0 0 0;
}

div.delicious ul li a {
  font-weight: normal;
}

div.delicious h2 a {
  color: #000;
}
</style>
<div id="itw" class="yui-d0 yui-skin-sam">

  <?php echo $header; ?>

  <yml:form method="POST" view="YahooFullView">
   <fieldset>
    <legend>Delicious Username</legend>
    <input type="text" name="username" />
    <input type="submit" value="Refresh" />
   </fieldset>
  </yml:form>

  <?php echo $content; ?>
</div>